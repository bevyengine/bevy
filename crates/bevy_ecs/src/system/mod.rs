mod commands;
mod into_system;
mod into_thread_local;
mod query;
#[allow(clippy::module_inception)]
mod system;
mod system_chaining;
mod system_param;

pub use commands::*;
pub use into_system::*;
pub use into_thread_local::*;
pub use query::*;
pub use system::*;
pub use system_chaining::*;
pub use system_param::*;
