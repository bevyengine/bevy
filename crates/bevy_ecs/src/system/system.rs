use crate::resource::Resources;
use bevy_hecs::{ArchetypeComponent, TypeAccess, World};
use std::{any::TypeId, borrow::Cow};

/// Determines the strategy used to run the `run_thread_local` function in a [System]
#[derive(Copy, Clone, Eq, PartialEq, Debug)]
pub enum ThreadLocalExecution {
    Immediate,
    NextFlush,
}

#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
pub struct SystemId(pub usize);

impl SystemId {
    #[allow(clippy::new_without_default)]
    pub fn new() -> Self {
        SystemId(rand::random::<usize>())
    }
}

/// An ECS system that can be added to a [Schedule](crate::Schedule)
pub trait System: Send + Sync {
    fn name(&self) -> Cow<'static, str>;
    fn id(&self) -> SystemId;
    fn update(&mut self, world: &World);
    fn archetype_component_access(&self) -> &TypeAccess<ArchetypeComponent>;
    fn resource_access(&self) -> &TypeAccess<TypeId>;
    fn thread_local_execution(&self) -> ThreadLocalExecution;
    fn run(&mut self, world: &World, resources: &Resources);
    fn run_thread_local(&mut self, world: &mut World, resources: &mut Resources);
    fn initialize(&mut self, _world: &mut World, _resources: &mut Resources) {}
}
