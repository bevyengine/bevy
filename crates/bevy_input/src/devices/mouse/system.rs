use crate::{
    core::{ElementState, Input},
    devices::mouse::{MouseButton, MouseButtonInput, MouseButtonInputState},
};
use bevy_app::prelude::{EventReader, Events};
use bevy_ecs::{Local, Res, ResMut};
use bevy_math::Vec2;

/// Updates the Input<MouseButton> resource with the latest MouseButtonInput events
pub fn mouse_button_input_system(
    mut state: Local<MouseButtonInputState>,
    mut mouse_button_input: ResMut<Input<MouseButton>>,
    mouse_button_input_events: Res<Events<MouseButtonInput>>,
) {
    mouse_button_input.update();
    for event in state
        .mouse_button_input_event_reader
        .iter(&mouse_button_input_events)
    {
        match event.state {
            ElementState::Pressed => mouse_button_input.press(event.button),
            ElementState::Released => mouse_button_input.release(event.button),
        }
    }
}
