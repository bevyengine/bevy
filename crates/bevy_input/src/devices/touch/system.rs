use bevy_input::devices::touch::{TouchSystemState, Touches, TouchInput, TouchPhase, Touch};

/// Updates the Touches resource with the latest TouchInput events
pub fn touch_screen_input_system(
    mut state: Local<TouchSystemState>,
    mut touch_state: ResMut<Touches>,
    touch_input_events: Res<Events<TouchInput>>,
) {
    let touch_state = &mut *touch_state;
    for released_id in touch_state.just_released.iter() {
        touch_state.active_touches.remove(&released_id);
    }

    for cancelled_id in touch_state.just_cancelled.iter() {
        touch_state.active_touches.remove(&cancelled_id);
    }

    touch_state.just_pressed.clear();
    touch_state.just_cancelled.clear();

    for event in state.touch_event_reader.iter(&touch_input_events) {
        let active_touch = touch_state.active_touches.get(&event.id);
        match event.phase {
            TouchPhase::Started => {
                touch_state.active_touches.insert(
                    event.id,
                    Touch {
                        id: event.id,
                        start_position: event.position,
                        previous_position: event.position,
                        position: event.position,
                    },
                );
                touch_state.just_pressed.insert(event.id);
            }
            TouchPhase::Moved => {
                let old_touch = active_touch.unwrap();
                let mut new_touch = old_touch.clone();
                new_touch.previous_position = new_touch.position;
                new_touch.position = event.position;
                touch_state.active_touches.insert(event.id, new_touch);
            }
            TouchPhase::Ended => {
                touch_state.just_released.insert(event.id);
            }
            TouchPhase::Cancelled => {
                touch_state.just_cancelled.insert(event.id);
            }
        };
    }
}

#[derive(Default)]
pub struct TouchSystemState {
    touch_event_reader: EventReader<TouchInput>,
}

