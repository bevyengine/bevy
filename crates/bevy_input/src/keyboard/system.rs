//! Input update system module

use crate::{
    core::{ElementState, Input},
    keyboard::{KeyCode, KeyboardInput, KeyboardInputState},
};
use bevy_app::prelude::*;
use bevy_ecs::{Local, Res, ResMut};

/// Updates the Input<KeyCode> resource with the latest KeyboardInput events
pub fn keyboard_input_system(
    mut state: Local<KeyboardInputState>,
    mut keyboard_input: ResMut<Input<KeyCode>>,
    keyboard_input_events: Res<Events<KeyboardInput>>,
) {
    keyboard_input.update();
    for event in state
        .keyboard_input_event_reader
        .iter(&keyboard_input_events)
    {
        if let KeyboardInput {
            key_code: Some(key_code),
            state,
            ..
        } = event
        {
            match state {
                ElementState::Pressed => keyboard_input.press(*key_code),
                ElementState::Released => keyboard_input.release(*key_code),
            }
        }
    }
}
