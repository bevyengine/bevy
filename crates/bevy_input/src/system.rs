use crate::{
    keyboard::{KeyCode, KeyboardInput},
    ElementState,
};
use bevy_app::{
    prelude::{EventReader, Events},
    AppExit,
};
use bevy_ecs::ResMut;

/// Sends the AppExit event whenever the "esc" key is pressed.
pub fn exit_on_esc_system(
    mut keyboard_input_events: EventReader<KeyboardInput>,
    mut app_exit_events: ResMut<Events<AppExit>>,
) {
    for event in keyboard_input_events.iter() {
        if let Some(key_code) = event.key_code {
            if event.state == ElementState::Pressed && key_code == KeyCode::Escape {
                app_exit_events.send(AppExit);
            }
        }
    }
}
