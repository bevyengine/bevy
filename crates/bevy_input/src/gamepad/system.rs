use crate::{
    gamepad::{
        GamepadAxis, GamepadButton, GamepadEvent, GamepadEventRaw, GamepadEventType,
        GamepadSettings, Gamepads, ALL_AXIS_TYPES, ALL_BUTTON_TYPES,
    },
    Axis, Input,
};
use bevy_app::{EventReader, EventWriter};
use bevy_ecs::system::{Res, ResMut};
use bevy_utils::tracing::info;

/// Monitors gamepad connection and disconnection events and updates the [`Gamepads`] resource accordingly.
///
/// ## Schedule
///
/// By default, this system runs during `CoreStage::PreUpdate` when added via the [`InputPlugin`](crate::InputPlugin).
///
/// ## Note
///
/// Whenever a [`Gamepad`](crate::gamepad::Gamepad) connects or disconnects, an information gets printed to the console using the [`info!`] macro.
pub fn gamepad_connection_system(
    mut gamepads: ResMut<Gamepads>,
    mut gamepad_event: EventReader<GamepadEvent>,
) {
    for event in gamepad_event.iter() {
        match event.event_type {
            GamepadEventType::Connected => {
                gamepads.register(event.gamepad);
                info!("{:?} Connected", event.gamepad);
            }
            GamepadEventType::Disconnected => {
                gamepads.deregister(&event.gamepad);
                info!("{:?} Disconnected", event.gamepad);
            }
            _ => (),
        }
    }
}

/// Modifies the gamepad resources and sends out gamepad events.
///
/// The resources [`Input<GamepadButton>`], [`Axis<GamepadAxis>`], and [`Axis<GamepadButton>`] are updated
/// and the [`GamepadEvent`]s are sent according to the received [`GamepadEventRaw`]s.
///
/// ## Differences
///
/// The main difference between the events and the resources is that the latter have convenient functions
/// like [`Input::pressed`], [`Input::just_pressed`], and [`Input::just_released`] or [`Axis::get`], and
/// [`Axis::set`].
pub fn gamepad_event_system(
    mut button_input: ResMut<Input<GamepadButton>>,
    mut axis: ResMut<Axis<GamepadAxis>>,
    mut button_axis: ResMut<Axis<GamepadButton>>,
    mut raw_events: EventReader<GamepadEventRaw>,
    mut events: EventWriter<GamepadEvent>,
    settings: Res<GamepadSettings>,
) {
    button_input.clear();
    for event in raw_events.iter() {
        match event.event_type {
            GamepadEventType::Connected => {
                events.send(GamepadEvent::new(event.gamepad, event.event_type.clone()));
                for button_type in ALL_BUTTON_TYPES.iter() {
                    let gamepad_button = GamepadButton::new(event.gamepad, *button_type);
                    button_input.reset(gamepad_button);
                    button_axis.set(gamepad_button, 0.0);
                }
                for axis_type in ALL_AXIS_TYPES.iter() {
                    axis.set(GamepadAxis::new(event.gamepad, *axis_type), 0.0);
                }
            }
            GamepadEventType::Disconnected => {
                events.send(GamepadEvent::new(event.gamepad, event.event_type.clone()));
                for button_type in ALL_BUTTON_TYPES.iter() {
                    let gamepad_button = GamepadButton::new(event.gamepad, *button_type);
                    button_input.reset(gamepad_button);
                    button_axis.remove(gamepad_button);
                }
                for axis_type in ALL_AXIS_TYPES.iter() {
                    axis.remove(GamepadAxis::new(event.gamepad, *axis_type));
                }
            }
            GamepadEventType::AxisChanged(axis_type, value) => {
                let gamepad_axis = GamepadAxis::new(event.gamepad, axis_type);
                if let Some(filtered_value) = settings
                    .get_axis_settings(gamepad_axis)
                    .filter(value, axis.get(gamepad_axis))
                {
                    axis.set(gamepad_axis, filtered_value);
                    events.send(GamepadEvent::new(
                        event.gamepad,
                        GamepadEventType::AxisChanged(axis_type, filtered_value),
                    ));
                }
            }
            GamepadEventType::ButtonChanged(button_type, value) => {
                let gamepad_button = GamepadButton::new(event.gamepad, button_type);
                if let Some(filtered_value) = settings
                    .get_button_axis_settings(gamepad_button)
                    .filter(value, button_axis.get(gamepad_button))
                {
                    button_axis.set(gamepad_button, filtered_value);
                    events.send(GamepadEvent::new(
                        event.gamepad,
                        GamepadEventType::ButtonChanged(button_type, filtered_value),
                    ))
                }

                let button_property = settings.get_button_settings(gamepad_button);
                if button_input.pressed(gamepad_button) {
                    if button_property.is_released(value) {
                        button_input.release(gamepad_button);
                    }
                } else if button_property.is_pressed(value) {
                    button_input.press(gamepad_button);
                }
            }
        }
    }
}
