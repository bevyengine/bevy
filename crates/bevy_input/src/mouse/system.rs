use crate::{
    mouse::{MouseButton, MouseButtonInput},
    ButtonState, Input,
};
use bevy_ecs::event::EventReader;
use bevy_ecs::system::ResMut;

/// Updates the [`Input<MouseButton>`] resource with the latest [`MouseButtonInput`] events.
///
/// ## Differences
///
/// The main difference between the [`MouseButtonInput`] event and the [`Input<MouseButton>`] resource is that
/// the latter has convenient functions like [`Input::pressed`], [`Input::just_pressed`] and [`Input::just_released`].
pub fn mouse_button_input_system(
    mut mouse_button_input: ResMut<Input<MouseButton>>,
    mut mouse_button_input_events: EventReader<MouseButtonInput>,
) {
    mouse_button_input.clear();
    for event in mouse_button_input_events.iter() {
        match event.state {
            ButtonState::Pressed => mouse_button_input.press(event.button),
            ButtonState::Released => mouse_button_input.release(event.button),
        }
    }
}
