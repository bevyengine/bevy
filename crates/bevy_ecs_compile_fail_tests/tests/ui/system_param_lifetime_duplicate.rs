use bevy_ecs::prelude::*;
use bevy_ecs::system::SystemParam;

#[allow(dead_code)]
#[derive(Debug, Resource)]
struct PlayerName(String);

// must fail to compile
#[allow(dead_code)]
#[derive(Debug, SystemParam)]
struct DuplicateWorldLifetime<'w, 'world> {
    player_name: Res<'w, PlayerName>,
}

// must fail to compile
#[allow(dead_code)]
#[derive(Debug, SystemParam)]
struct DuplicateStateLifetime<'s, 'cache> {
    player_name: Local<'cache, String>,
}
