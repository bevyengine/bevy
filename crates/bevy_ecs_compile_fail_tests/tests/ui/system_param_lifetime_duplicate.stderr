error: invalid lifetime name: already using "'w" for 'world lifetime;
                               synonyms for 'world are 'w, 'world
  --> tests/ui/system_param_lifetime_duplicate.rs:11:31
   |
11 | struct DuplicateWorldLifetime<'w, 'world> {
   |                               ^^

error: invalid lifetime name: already using "'s" for 'state lifetime;
                               synonyms for 'state are 's, 'state, 'param, 'cache, 'query
  --> tests/ui/system_param_lifetime_duplicate.rs:18:31
   |
18 | struct DuplicateStateLifetime<'s, 'cache> {
   |                               ^^

error[E0601]: `main` function not found in crate `$CRATE`
  --> tests/ui/system_param_lifetime_duplicate.rs:20:2
   |
20 | }
   |  ^ consider adding a `main` function to `$DIR/tests/ui/system_param_lifetime_duplicate.rs`

error[E0392]: parameter `'world` is never used
  --> tests/ui/system_param_lifetime_duplicate.rs:11:35
   |
11 | struct DuplicateWorldLifetime<'w, 'world> {
   |                                   ^^^^^^ unused parameter
   |
   = help: consider removing `'world`, referring to it in a field, or using a marker such as `PhantomData`

error[E0392]: parameter `'s` is never used
  --> tests/ui/system_param_lifetime_duplicate.rs:18:31
   |
18 | struct DuplicateStateLifetime<'s, 'cache> {
   |                               ^^ unused parameter
   |
   = help: consider removing `'s`, referring to it in a field, or using a marker such as `PhantomData`
